microdnf-plugins/changelog_plugin/changelog.cpp

```diff
@@ -0,0 +1,228 @@
+the Free Software Foundation, either version 2 of the License, or
+(at your option) any later version.
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+
+#include "changelog.hpp"
+
+#include "microdnf/context.hpp"
+
+#include <libdnf/conf/option_string.hpp>
+#include <libdnf/rpm/package.hpp>
+#include <libdnf/rpm/package_query.hpp>
+#include <libdnf/rpm/package_set.hpp>
+
+#include <ctime>
+#include <iomanip>
+#include <iostream>
+#include <sstream>
+
+
+namespace microdnf {
+
+
+using namespace libdnf::cli;
+
+
+ChangelogCommand::ChangelogCommand(Command & parent) : Command(parent, "changelog") {
+    auto & ctx = static_cast<Context &>(get_session());
+    auto & parser = ctx.get_argument_parser();
+
+    auto & cmd = *get_argument_parser_command();
+    cmd.set_short_description("Show changelog data of packages");
+
+    since_option = dynamic_cast<libdnf::OptionNumber<std::int64_t> *>(
+        parser.add_init_value(std::unique_ptr<libdnf::OptionNumber<std::int64_t>>(
+            new libdnf::OptionNumber<int64_t>(0, [](const std::string & value) {
+                struct tm time_m;
+                std::istringstream ss(value);
+                ss >> std::get_time(&time_m, "%Y-%m-%d");
+                if (ss.fail()) {
+                    throw std::runtime_error(fmt::format("Invalid date: {}", value));
+                }
+                return static_cast<int64_t>(mktime(&time_m));
+            }))));
+
+    count_option = dynamic_cast<libdnf::OptionNumber<std::int32_t> *>(
+        parser.add_init_value(std::unique_ptr<libdnf::OptionNumber<std::int32_t>>(new libdnf::OptionNumber<int>(0))));
+
+    upgrades_option = dynamic_cast<libdnf::OptionBool *>(
+        parser.add_init_value(std::unique_ptr<libdnf::OptionBool>(new libdnf::OptionBool(false))));
+
+    auto since = parser.add_new_named_arg("since");
+    since->set_long_name("since");
+    since->set_short_description(
+        "show changelog entries since DATE. To avoid ambiguosity, YYYY-MM-DD format is required.");
+    since->set_has_value(true);
+    since->link_value(since_option);
+
+    auto count = parser.add_new_named_arg("count");
+    count->set_long_name("count");
+    count->set_short_description("show given number of changelog entries per package");
+    count->set_has_value(true);
+    count->link_value(count_option);
+
+    auto upgrades = parser.add_new_named_arg("upgrades");
+    upgrades->set_long_name("upgrades");
+    upgrades->set_short_description(
+        "show only new changelog entries for packages, that provide an upgrade for some of already installed "
+        "packages.");
```

Comment:
> show new changelog entries for packages that provide an upgrade for an already installed package

Reply:
The text was taken from DNF4. I will change it.